File: java/com/epicdragonworld/gameserver/network/packets/receivable/AccountAuthenticationRequest.java
/*
 * This file is part of the Epic Dragon World project.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package com.epicdragonworld.gameserver.network.packets.receivable;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.logging.Logger;

import com.epicdragonworld.gameserver.managers.DatabaseManager;
import com.epicdragonworld.gameserver.network.GameClient;
import com.epicdragonworld.gameserver.network.ReceivablePacket;
import com.epicdragonworld.gameserver.network.packets.sendable.AccountAuthenticationResult;

/**
 * @author Pantelis Andrianakis
 */
public class AccountAuthenticationRequest
{
	private static final Logger LOGGER = Logger.getLogger(AccountAuthenticationRequest.class.getName());
	
	private static final String ACCOUNT_QUERY = "SELECT * FROM accounts WHERE account=?";
	private static final String ACCOUNT_INFO_UPDATE = "UPDATE accounts SET last_active=?, last_ip=? WHERE account=?";
	private static final int STATUS_NOT_FOUND = 0;
	private static final int STATUS_WRONG_PASSWORD = 3;
	@SuppressWarnings("unused")
	private static final int STATUS_TOO_MANY_ONLINE = 4; // TODO: STATUS_TOO_MANY_ONLINE
	private static final int STATUS_AUTHENTICATED = 100;
	
	public AccountAuthenticationRequest(GameClient client, ReceivablePacket packet)
	{
		// Read data.
		final String accountName = packet.readString().toLowerCase();
		final String passwordHash = packet.readString();
		
		// Local data.
		String storedPassword = "";
		int status = STATUS_NOT_FOUND;
		
		// Get data from database.
		try (Connection con = DatabaseManager.getInstance().getConnection();
			PreparedStatement ps = con.prepareStatement(ACCOUNT_QUERY))
		{
			ps.setString(1, accountName);
			try (ResultSet rset = ps.executeQuery())
			{
				while (rset.next())
				{
					storedPassword = rset.getString("password");
					status = rset.getInt("status");
				}
			}
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
		
		// Account status issue.
		// 0 does not exist, 1 banned, 2 requires activation, 3 wrong password, 4 too many online, 5 authenticated
		if (status < STATUS_WRONG_PASSWORD)
		{
			client.channelSend(new AccountAuthenticationResult(status));
			return;
		}
		
		// Wrong password.
		if (!passwordHash.equals(storedPassword))
		{
			client.channelSend(new AccountAuthenticationResult(STATUS_WRONG_PASSWORD));
			return;
		}
		
		// TODO: STATUS_TOO_MANY_ONLINE
		
		// Authentication was successful.
		client.setAccountName(accountName);
		client.channelSend(new AccountAuthenticationResult(STATUS_AUTHENTICATED));
		
		// Update last login date and IP address.
		try (Connection con = DatabaseManager.getInstance().getConnection();
			PreparedStatement ps = con.prepareStatement(ACCOUNT_INFO_UPDATE))
		{
			ps.setLong(1, System.currentTimeMillis());
			ps.setString(2, client.getIp());
			ps.setString(3, accountName);
			ps.execute();
		}
		catch (Exception e)
		{
			LOGGER.warning(e.getMessage());
		}
	}
}

File: EpicDragonWorldServer/Network/Packets/Receivable/AccountAuthenticationRequest.cs
using MySql.Data.MySqlClient;
using System;

/**
 * Author: Pantelis Andrianakis
 * Date: November 7th 2018
 */
public class AccountAuthenticationRequest
{
    private static readonly string ACCOUNT_INFO_QUERY = "SELECT * FROM accounts WHERE account=@account";
    private static readonly string ACCOUNT_INFO_UPDATE_QUERY = "UPDATE accounts SET last_active=@last_active, last_ip=@ip WHERE account=@account";
    private static readonly string ACCOUNT_CREATE_QUERY = "INSERT INTO accounts (account, password, status) values (@account, @password, 3)";
    private static readonly int STATUS_NOT_FOUND = 0;
    private static readonly int STATUS_WRONG_PASSWORD = 3;
    private static readonly int STATUS_ALREADY_ONLINE = 4;
    private static readonly int STATUS_TOO_MANY_ONLINE = 5;
    private static readonly int STATUS_INCORRECT_CLIENT = 6;
    private static readonly int STATUS_AUTHENTICATED = 100;

    public AccountAuthenticationRequest(GameClient client, ReceivablePacket packet)
    {
        // Read data.
        double clientVersion = packet.ReadDouble();
        string accountName = packet.ReadString().ToLowerInvariant();
        string passwordHash = packet.ReadString();

        // Client version check.
        if (clientVersion != Config.CLIENT_VERSION)
        {
            client.ChannelSend(new AccountAuthenticationResult(STATUS_INCORRECT_CLIENT));
            return;
        }

        // Replace illegal characters.
        foreach (char c in Util.ILLEGAL_CHARACTERS)
        {
            accountName = accountName.Replace(c, '\'');
        }

        // Account name checks.
        if ((accountName.Length < 2) || (accountName.Length > 20) || accountName.Contains("'") || (passwordHash.Length == 0)) // 20 should not happen, checking it here in case of client cheat.
        {
            client.ChannelSend(new AccountAuthenticationResult(STATUS_NOT_FOUND));
            return;
        }

        // Get data from database.
        string storedPassword = "";
        int status = STATUS_NOT_FOUND;
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(ACCOUNT_INFO_QUERY, con);
            cmd.Parameters.AddWithValue("account", accountName);
            MySqlDataReader reader = cmd.ExecuteReader();
            while (reader.Read())
            {
                storedPassword = reader.GetString("password");
                status = reader.GetInt32("status");
            }
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }

        // In case of auto create accounts configuration.
        if ((status == 0) && Config.ACCOUNT_AUTO_CREATE)
        {
            // Create account.
            try
            {
                MySqlConnection con = DatabaseManager.GetConnection();
                MySqlCommand cmd = new MySqlCommand(ACCOUNT_CREATE_QUERY, con);
                cmd.Parameters.AddWithValue("account", accountName);
                cmd.Parameters.AddWithValue("password", passwordHash);
                cmd.ExecuteNonQuery();
                con.Close();
            }
            catch (Exception e)
            {
                LogManager.Log(e.ToString());
            }
            LogManager.Log("Created account " + accountName + ".");
        }
        else // Account status issue.
        {
            // 0 does not exist, 1 banned, 2 requires activation, 3 wrong password, 4 too many online, 100 authenticated
            if (status < STATUS_WRONG_PASSWORD)
            {
                client.ChannelSend(new AccountAuthenticationResult(status));
                return;
            }

            // Wrong password.
            if (!passwordHash.Equals(storedPassword))
            {
                client.ChannelSend(new AccountAuthenticationResult(STATUS_WRONG_PASSWORD));
                return;
            }
        }

        // Kick existing logged client.
        GameClient existingClient = WorldManager.GetClientByAccountName(accountName);
        if (existingClient != null)
        {
            existingClient.ChannelSend(new Logout());
            WorldManager.RemoveClient(existingClient);
            client.ChannelSend(new AccountAuthenticationResult(STATUS_ALREADY_ONLINE));
            return;
        }

        // Too many online users.
        if (WorldManager.GetOnlineCount() >= Config.MAXIMUM_ONLINE_USERS)
        {
            client.ChannelSend(new AccountAuthenticationResult(STATUS_TOO_MANY_ONLINE));
            return;
        }

        // Authentication was successful.
        WorldManager.AddClient(client);
        client.SetAccountName(accountName);
        client.ChannelSend(new AccountAuthenticationResult(STATUS_AUTHENTICATED));

        // Update last login date and IP address.
        try
        {
            MySqlConnection con = DatabaseManager.GetConnection();
            MySqlCommand cmd = new MySqlCommand(ACCOUNT_INFO_UPDATE_QUERY, con);
            cmd.Parameters.AddWithValue("last_active", DateTimeOffset.Now.ToUnixTimeSeconds());
            cmd.Parameters.AddWithValue("ip", client.GetIp());
            cmd.Parameters.AddWithValue("account", accountName);
            cmd.ExecuteNonQuery();
            con.Close();
        }
        catch (Exception e)
        {
            LogManager.Log(e.ToString());
        }
    }
}
